src/include/storage/sql_table.h

```diff
@@ -27,6 +31,110 @@ class SqlTable {
     DataTable *data_table;
     BlockLayout layout;
     ColumnMap column_map;
+    InverseColumnMap inverse_column_map;
+    // TODO(John): Add 'default_value_map' (dynamic) for col_oid->default_val
+  };
+
+  /**
+   * Iterator for all the slots, claimed or otherwise, in the data table. This is useful for sequential scans.
+   */
+  class SlotIterator {
+    // TODO(Yashwanth): Slot iterator currently flawed, for a scan on a certain version, it MUST begin on the latest
+    // version it sees
+   public:
+    /**
+     * @return reference to the underlying tuple slot
+     */
+    const TupleSlot &operator*() const { return *current_it_; }
+
+    /**
+     * @return pointer to the underlying tuple slot
+     */
+    const TupleSlot *operator->() const { return &(*current_it_); }
+
+    /**
+     * pre-fix increment.
+     * @return self-reference after the iterator is advanced
+     */
+    SlotIterator &operator++() {
+      TERRIER_ASSERT(!is_end_, "Cannot increment an end iterator");
+      current_it_++;
+      AdvanceOnEndOfDatatable_();
+      return *this;
+    }
+
+    /**
+     * post-fix increment.
+     * @return copy of the iterator equal to this before increment
+     */
+    const SlotIterator operator++(int) {
+      SlotIterator copy = *this;
+      operator++();
+      return copy;
+    }
+
+    /**
+     * Equality check.  Either both iterators are "end" or they must match
+     * on both their observed version and the underlying tuple.
+     * @param other other iterator to compare to
+     * @return if the two iterators point to the same slotcolumn_ids
+     */
+    bool operator==(const SlotIterator &other) const {
+      // First "is_end" check is that both are end, the second protects the iterator check through short-circuit
+      return (is_end_ && other.is_end_) ||
+             (is_end_ == other.is_end_ && txn_version_ == other.txn_version_ && current_it_ == other.current_it_);
+    }
+
+    /**
+     * Inequality check.
+     * @param other other iterator to compare to
+     * @return if the two iterators are not equal
+     */
+    bool operator!=(const SlotIterator &other) const { return !this->operator==(other); }
+
+    DataTable::SlotIterator *GetDataTableSlotIterator() { return &current_it_; }
+
+   private:
+    friend class SqlTable;
+
+    SlotIterator(const common::ConcurrentMap<layout_version_t, DataTableVersion> *tables,
+                 const layout_version_t txn_version, bool is_end)
```

Comment:
How can a running transaction know which 'txn_version' to use in its scans? Your implementation in the AdvanceOnEndOfDatatable_() suggests that layout versions are incremented sequentially, so the txn cannot simply pass its txn_id here.

Reply:
A transaction has a single `txn_version` that is constant across its entire lifetime and is determined by what is visible in the catalog when the binder executes.  `AdvanceOnEndOfDatabase_` detects when we've iterated over all of the tuples in a data table and advances to the next data table (newest to oldest) when necessary.